File: src/config_fixer.py
import yaml
import logging
from pathlib import Path
from typing import Any, Dict, List

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

DEFAULT_IMAGE = "nginx:latest"
DEFAULT_RESOURCES = {
    "limits": {"cpu": "500m", "memory": "256Mi"},
    "requests": {"cpu": "250m", "memory": "128Mi"},
}
DEFAULT_ENV = {"name": "ENV", "value": "production"}

def load_yaml(path: Path) -> Dict[str, Any]:
    with path.open("r", encoding="utf-8") as f:
        return yaml.safe_load(f) or {}

def dump_yaml(data: Dict[str, Any], path: Path) -> None:
    with path.open("w", encoding="utf-8") as f:
        yaml.safe_dump(data, f, sort_keys=False)

def ensure_container_defaults(container: Dict[str, Any]) -> None:
    if "image" not in container:
        container["image"] = DEFAULT_IMAGE

    resources = container.setdefault("resources", {})
    limits = resources.setdefault("limits", {})
    requests = resources.setdefault("requests", {})
    for key in ("cpu", "memory"):
        limits.setdefault(key, DEFAULT_RESOURCES["limits"][key])
        requests.setdefault(key, DEFAULT_RESOURCES["requests"][key])

    env_list: List[Dict[str, str]] = container.setdefault("env", [])
    if not any(env.get("name") == DEFAULT_ENV["name"] for env in env_list):
        env_list.append(DEFAULT_ENV)

def apply_fixes(manifest: Dict[str, Any]) -> None:
    try:
        containers = manifest["spec"]["template"]["spec"]["containers"]
    except (KeyError, TypeError):
        logger.warning("Not a pod template; nothing to fix.")
        return

    if not isinstance(containers, list):
        return

    for container in containers:
        if isinstance(container, dict):
            ensure_container_defaults(container)

def main(argv: List[str]) -> None:
    if len(argv) != 2:
        print("Usage: python config_fixer.py <path-to-manifest.yaml>")
        sys.exit(1)

    path = Path(argv[1])
    if not path.is_file():
        print(f"Error: {path} does not exist or is not a file.")
        sys.exit(1)

    manifest = load_yaml(path)
    apply_fixes(manifest)
    dump_yaml(manifest, path)
    logger.info(f"Fixed configuration written to {path}")

if __name__ == "__main__":
    import sys
    main(sys.argv)

File: src/config_fixer.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"

	"gopkg.in/yaml.v3"
)

const (
	defaultImage = "nginx:latest"
)

var defaultResources = map[string]map[string]string{
	"limits":   {"cpu": "500m", "memory": "256Mi"},
	"requests": {"cpu": "250m", "memory": "128Mi"},
}

var defaultEnv = map[string]string{
	"name":  "ENV",
	"value": "production",
}

type Container struct {
	Image     string                 `yaml:"image,omitempty"`
	Resources map[string]map[string]string `yaml:"resources,omitempty"`
	Env       []map[string]string    `yaml:"env,omitempty"`
}

type PodSpec struct {
	Containers []Container `yaml:"containers"`
}

type PodTemplateSpec struct {
	Spec PodSpec `yaml:"spec"`
}

type PodSpecWrapper struct {
	Spec PodTemplateSpec `yaml:"spec"`
}

func ensureContainerDefaults(c *Container) {
	if c.Image == "" {
		c.Image = defaultImage
	}

	if c.Resources == nil {
		c.Resources = make(map[string]map[string]string)
	}
	for key, val := range defaultResources {
		if c.Resources[key] == nil {
			c.Resources[key] = make(map[string]string)
		}
		for k, v := range val {
			c.Resources[key][k] = v
		}
	}

	if c.Env == nil {
		c.Env = []map[string]string{}
	}
	found := false
	for _, env := range c.Env {
		if env["name"] == defaultEnv["name"] {
			found = true
			break
		}
	}
	if !found {
		c.Env = append(c.Env, defaultEnv)
	}
}

func applyFixes(manifest *PodSpecWrapper) {
	for i := range manifest.Spec.Spec.Containers {
		ensureContainerDefaults(&manifest.Spec.Spec.Containers[i])
	}
}

func main() {
	if len(os.Args) != 2 {
		fmt.Println("Usage: go run config_fixer.go <path-to-manifest.yaml>")
		os.Exit(1)
	}

	path := os.Args[1]
	data, err := ioutil.ReadFile(path)
	if err != nil {
		log.Fatalf("Failed to read file: %v", err)
	}

	var manifest PodSpecWrapper
	err = yaml.Unmarshal(data, &manifest)
	if err != nil {
		log.Fatalf("Failed to unmarshal YAML: %v", err)
	}

	applyFixes(&manifest)

	data, err = yaml.Marshal(&manifest)
	if err != nil {
		log.Fatalf("Failed to marshal YAML: %v", err)
	}

	err = ioutil.WriteFile(path, data, 0644)
	if err != nil {
		log.Fatalf("Failed to write file: %v", err)
	}

	log.Printf("Fixed configuration written to %s", path)
}